test_conformance/subgroups/subhelpers.h

```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
+    static const char *add_typedef() { return "typedef long16 Type;\n"; }
+    using scalar_type = cl_long;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return int64_ok(device);
+    }
+};
+// cl_ulong
+template <> struct TypeManager<cl_ulong> : public CommonTypeManager<cl_ulong>
+{
+    static const char *name() { return "ulong"; }
+    static const char *add_typedef() { return "typedef ulong Type;\n"; }
+    static const bool type_supported(cl_device_id device)
+    {
+        return int64_ok(device);
+    }
+};
+template <> struct TypeManager<cl_ulong2> : public CommonTypeManager<cl_ulong2>
+{
+    static const char *name() { return "ulong2"; }
+    static const char *add_typedef() { return "typedef ulong2 Type;\n"; }
+    using scalar_type = cl_ulong;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return int64_ok(device);
+    }
+};
+template <>
+struct TypeManager<subgroups::cl_ulong3>
+    : public CommonTypeManager<subgroups::cl_ulong3>
+{
+    static const char *name() { return "ulong3"; }
+    static const char *add_typedef() { return "typedef ulong3 Type;\n"; }
+    typedef std::true_type is_sb_vector_size3;
+    using scalar_type = cl_ulong;
+    static const bool type_supported(cl_device_id device)
+    {
+        return int64_ok(device);
+    }
+};
+template <> struct TypeManager<cl_ulong4> : public CommonTypeManager<cl_ulong4>
+{
+    static const char *name() { return "ulong4"; }
+    static const char *add_typedef() { return "typedef ulong4 Type;\n"; }
+    using scalar_type = cl_ulong;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return int64_ok(device);
+    }
+};
+template <> struct TypeManager<cl_ulong8> : public CommonTypeManager<cl_ulong8>
+{
+    static const char *name() { return "ulong8"; }
+    static const char *add_typedef() { return "typedef ulong8 Type;\n"; }
+    using scalar_type = cl_ulong;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return int64_ok(device);
+    }
+};
+template <>
+struct TypeManager<cl_ulong16> : public CommonTypeManager<cl_ulong16>
+{
+    static const char *name() { return "ulong16"; }
+    static const char *add_typedef() { return "typedef ulong16 Type;\n"; }
+    using scalar_type = cl_ulong;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return int64_ok(device);
+    }
 };
 
-
-template <> struct TypeIdentity<float, 0>
+// cl_float
+template <> struct TypeManager<cl_float> : public CommonTypeManager<cl_float>
+{
+    static const char *name() { return "float"; }
+    static const char *add_typedef() { return "typedef float Type;\n"; }
+    static cl_float identify_limits(ArithmeticOp operation)
```

Comment:
I was really baffled by what these functions do until I found where the spec talks about the identity _I_ for various operations. What does "identify_limits" refer to? For me something like get_identity_value would be much more clear